FirebaseFunctions/Tests/Unit/FunctionsSerializerTests.swift

```diff
@@ -213,12 +245,78 @@ class FunctionsSerializerTests: XCTestCase {
     XCTAssertEqual(expected, try serializer.encode(input) as? NSDictionary)
   }
 
-  func testDecodeMap() {
+  func testEncodeDictionaryWithInvalidElements() {
+    let input = ["TEST_CustomObj": CustomObject()] as NSDictionary
+
+    do {
+      let _ = try serializer.encode(input)
+      XCTFail("Expected an error")
+    } catch {
+      guard case let .unsupportedType(typeName: typeName) = error as? FunctionsSerializer
+        .Error else {
+        return XCTFail("Unexpected error: \(error)")
+      }
+
+      XCTAssertEqual(typeName, "CustomObject")
+    }
+  }
+
+  func testEncodeDictionaryWithInvalidNestedDictionary() {
+    let input =
+      ["TEST_NestedDict": ["TEST_CustomObj": CustomObject()] as NSDictionary] as NSDictionary
+
+    do {
+      let _ = try serializer.encode(input)
+      XCTFail("Expected an error")
+    } catch {
+      guard case let .unsupportedType(typeName: typeName) = error as? FunctionsSerializer
+        .Error else {
+        return XCTFail("Unexpected error: \(error)")
+      }
+
+      XCTAssertEqual(typeName, "CustomObject")
+    }
+  }
+
+  func testDecodeDictionary() throws {
     let input = ["foo": 1, "bar": "hello", "baz": [3, 9_876_543_210]] as NSDictionary
     let expected = ["foo": 1, "bar": "hello", "baz": [3, 9_876_543_210]] as NSDictionary
     XCTAssertEqual(expected, try serializer.decode(input) as? NSDictionary)
   }
 
+  func testDecodeDictionaryWithInvalidElements() {
+    let input = ["TEST_CustomObj": CustomObject()] as NSDictionary
+
+    do {
+      let _ = try serializer.decode(input)
+      XCTFail("Expected an error")
+    } catch {
+      guard case let .unsupportedType(typeName: typeName) = error as? FunctionsSerializer
+        .Error else {
+        return XCTFail("Unexpected error: \(error)")
+      }
+
+      XCTAssertEqual(typeName, "CustomObject")
+    }
+  }
+
+  func testDecodeDictionaryWithInvalidNestedDictionary() {
+    let input =
+      ["TEST_NestedDict": ["TEST_CustomObj": CustomObject()] as NSDictionary] as NSDictionary
+
+    do {
+      let _ = try serializer.decode(input)
+      XCTFail("Expected an error")
+    } catch {
+      guard case let .unsupportedType(typeName: typeName) = error as? FunctionsSerializer
+        .Error else {
+        return XCTFail("Unexpected error: \(error)")
+      }
+
+      XCTAssertEqual(typeName, "CustomObject")
+    }
```

Comment:
I think this pattern used here and elsewhere may be expressed in something more concise like:
```suggestion
    XCTAssertThrowsError(try serializer.decode(input)) { (error) in
        guard case let .unsupportedType(typeName: typeName) = error as? FunctionsSerializer
                .Error else {
                return XCTFail("Unexpected error: \(error)")
              }
        XCTAssertEqual(typeName, "CustomObject")
    }
```

Reply:
You’re right, `XCTAssertThrowsError` is probably a better choice for non-`async` functions.